molr-mole-core/src/main/java/io/molr/mole/core/support/OngoingMissionRun.java

```diff
@@ -41,4 +44,107 @@ public OngoingMissionRun and() {
                 .map(MissionState::result);
     }
 
+    public Result whenFinished(Duration timeout) {
+        return awaitFinished().block(timeout);
+    }
+
+    public Result whenFinished() {
+        return awaitFinished().block();
+    }
+
+    public Mono<MissionHandle> asyncHandle() {
+        return handle;
+    }
+
+    public MissionHandle awaitHandle() {
+        return handle.block();
+    }
+
+    public void instruct(StrandCommand command) {
+        handle.subscribe(h -> mole.instructRoot(h, command));
+    }
+
+    public void await(RunStateValidator validator) {
+        filterState(validator, handle, mole)
+                .blockFirst();
+    }
+
+    public void await(RunStateValidator validator, Duration timeout) {
+        filterState(validator, handle, mole)
+                .blockFirst(timeout);
+    }
+
+    private static Flux<MissionState> filterState(RunStateValidator validator, Mono<MissionHandle> handle, Mole mole) {
+        return handle
+                .flatMapMany(mole::statesFor)
+                .filter(s -> validator.test(s.runState()));
+    }
+
+    public ReturnHelper returnResult() {
+        return new ReturnHelper<MissionState, Result>(Result.class);
+    }
+
+    public ReturnHelper returnState() {
+        return new ReturnHelper<MissionState, MissionState>(MissionState.class);
+    }
+
+    public static RunStateValidator runState(RunState runState) {
+        return RunStateValidator.validateState(runState);
+    }
+
+    public class ReturnHelper<T, R> {
+        private Function<T, R> function;
+
+        public ReturnHelper(Class<R> type) {
+            requireNonNull(type);
+            if (MissionState.class.equals(type)) {
+                function = new ReturnState<T, R>();
+            } else if (Result.class.equals(type)) {
+                function = new ReturnResult<T, R>();
+            }
+        }
+
+        public R when(RunStateValidator validator) {
+            return (R) returnValue(filterState(validator, handle, mole).blockFirst());
+        }
+
+        public R when(RunStateValidator validator, Duration timeout) {
+            return (R) returnValue(filterState(validator, handle, mole).blockFirst(timeout));
+        }
+
+        public R whenFinished() {
+            return when(runState(RunState.FINISHED));
+        }
+
+        public R whenFinished(Duration timeout) {
+            return when(runState(RunState.FINISHED), timeout);
+        }
+
+        public R whenPaused() {
+            return when(runState(RunState.PAUSED));
+        }
+
+        public R whenPaused(Duration timeout) {
+            return when(runState(RunState.PAUSED), timeout);
+        }
+
+        private R returnValue(MissionState missionState) {
+            return (R) function.apply((T) missionState);
+        }
+    }
+
+    private class ReturnState<T, R> implements Function<T, R> {
```

Comment:
I think there you need no type parameters .. this looks more like the identity function ...`Function.identity()`, or?